src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -0,0 +1,569 @@
+	     << ", last_committed_record_base="<< header.last_committed_record_base
+	     << ", written_to=" << header.written_to
+	     << ", flsg=" << header.flag
+	     << ", csum_type=" << header.csum_type
+	     << ", csum=" << header.csum
+	     << ", start=" << header.start
+	     << ", end=" << header.end
+             << ")";
+}
+
+
+CircularBoundedJournal::CircularBoundedJournal(NVMeBlockDevice* device,
+    const std::string path)
+  : device(device), path(path) {}
+
+CircularBoundedJournal::mkfs_ret
+CircularBoundedJournal::mkfs(mkfs_config_t& config)
+{
+  LOG_PREFIX(CircularBoundedJournal::mkfs);
+  return _open_device(path
+  ).safe_then([this, &config, FNAME]() -> mkfs_ret {
+    rbm_abs_addr start_addr = convert_paddr_to_abs_addr(
+      config.start);
+    assert(config.block_size == device->get_block_size());
+    ceph::bufferlist bl;
+    CircularBoundedJournal::cbj_header_t head;
+    head.magic = CBJOURNAL_MAGIC;
+    head.uuid = uuid_d(); // TODO
+    head.block_size = config.block_size;
+    rbm_abs_addr end_addr = convert_paddr_to_abs_addr(
+      config.end);
+    head.size = end_addr - start_addr
+      - device->get_block_size();
+    head.used_size = 0;
+    head.error = 0;
+    head.start_offset = device->get_block_size();
+    head.last_committed_record_base = 0;
+    head.written_to = head.start_offset;
+    head.applied_to = 0;
+    head.flag = 0;
+    head.csum_type = 0;
+    head.csum = 0;
+    head.cur_segment_seq = 0;
+    head.start = start_addr;
+    head.end = end_addr;
+    head.device_id = config.device_id;
+    encode(head, bl);
+    header = head;
+    DEBUG(
+      "initialize header block in CircularBoundedJournal, length {}",
+      bl.length());
+    return device_write_bl(start_addr, bl
+    ).handle_error(
+      mkfs_ertr::pass_further{},
+      crimson::ct_error::assert_all{
+      "Invalid error device_write during CircularBoundedJournal::mkfs"
+    }).safe_then([]() {
+      return mkfs_ertr::now();
+    });
+  }).handle_error(
+    mkfs_ertr::pass_further{},
+    crimson::ct_error::assert_all{
+    "Invalid error _open_device in CircularBoundedJournal::mkfs"
+  }).finally([this] {
+    if (device) {
+      return device->close();
+    } else {
+      return seastar::now();
+    }
+  });
+}
+
+CircularBoundedJournal::open_for_write_ertr::future<>
+CircularBoundedJournal::_open_device(const std::string path)
+{
+  ceph_assert(device);
+  return device->open(path, seastar::open_flags::rw
+  ).handle_error(
+    open_for_write_ertr::pass_further{},
+    crimson::ct_error::assert_all{
+      "Invalid error device->open"
+    }
+  );
+}
+
+ceph::bufferlist CircularBoundedJournal::encode_super()
+{
+  bufferlist bl;
+  encode(header, bl);
+  return bl;
+}
+
+CircularBoundedJournal::open_for_write_ret CircularBoundedJournal::open_for_write()
+{
+  return open_for_write(CBJOURNAL_START_ADDRESS);
+}
+
+CircularBoundedJournal::close_ertr::future<> CircularBoundedJournal::close()
+{
+  return write_super(
```

Comment:
Under what conditions does the super block actually change?  If it actually does change, is it ok to defer updating it until close?